tmexp/create_bow.py

```diff
@@ -0,0 +1,209 @@
+from collections import Counter
+import copy
+import logging
+import os
+import pickle
+from typing import Any, Dict, List, Optional
+
+import tqdm
+
+from .utils import (
+    check_exists,
+    check_remove_file,
+    create_directory,
+    create_language_list,
+)
+
+DIFF_MODEL = "diff"
+HALL_MODEL = "hall"
+SEP = ":"
+
+
+def check_fraction(fraction: float, arg_name: str) -> None:
+    if not (0 <= fraction <= 1):
+        raise RuntimeError(
+            "Argument '%s' must be in the range [0, 1], aborting." % arg_name
+        )
+
+
+def create_bow(
+    input_path: str,
+    output_dir: str,
+    dataset_name: Optional[str],
+    langs: Optional[List[str]],
+    exclude_langs: Optional[List[str]],
+    features: List[str],
+    force: bool,
+    topic_model: str,
+    min_word_frac: float,
+    max_word_frac: float,
+    log_level: str,
+) -> None:
+    logger = logging.getLogger(__name__)
+    logger.addHandler(logging.StreamHandler())
+    logger.setLevel(log_level)
+
+    check_exists(input_path)
+    create_directory(output_dir, logger)
+    if dataset_name is None:
+        dataset_name = topic_model
+    words_output_path = os.path.join(output_dir, "vocab." + dataset_name + ".txt")
+    check_remove_file(words_output_path, logger, force)
+    docword_output_path = os.path.join(output_dir, "docword." + dataset_name + ".txt")
+    check_remove_file(docword_output_path, logger, force)
+    docs_output_path = os.path.join(output_dir, "docs." + dataset_name + ".txt")
+    check_remove_file(docs_output_path, logger, force)
+    check_fraction(min_word_frac, "min-word-frac")
+    check_fraction(max_word_frac, "max-word-frac")
+
+    logger.info("Reading pickled dict from '%s' ..." % input_path)
+    with open(input_path, "rb") as fin:
+        input_dict = pickle.load(fin)
+
+    logger.info("Computing bag of words ...")
+    langs = create_language_list(langs, exclude_langs)
+    all_refs = sorted(input_dict["files_info"])
```

Comment:
After having reviewed the logic, I think this might be too simplistic: for the following logic to work, the refs should be sorted in creation time order, but for pytorch, we obtain:
```
v0.1.1
v0.1.10
v0.1.11
v0.1.12
v0.1.2
v0.1.3
v0.1.4
v0.1.5
v0.1.6
v0.1.7
v0.1.8
v0.1.9
v0.2.0
v0.3.0
v0.3.1
v0.4.0
v0.4.1
v1.0.0
v1.0.1
v1.0rc0
v1.0rc1
v1.1.0
```
We can see that this order is wrong: v0.1.{10,11,12} come before v0.1.2, and v1.0rc0 comes after v1.0.0 and even v1.0.1. I think we need to retrieve a date from gitbase and sort on that key instead.

Reply:
Yes my bad, adding this to `preprocess`. I will add an entry to the dictionnary:
{`files_info`: ...
 `files_content`: ...
  `ref_info`: [ref_1, ref_2 ... ref_n]}